File: packages/types/pulse.ts
import type { UUID, ISODate } from "./envelope";

export type TaskStatus = "todo" | "in_progress" | "review" | "done" | "blocked";
export type Priority = "low" | "normal" | "high" | "urgent";
export type Phase = "design" | "permit" | "construction" | "closeout";
export type MilestoneStatus = "upcoming" | "achieved" | "missed";
export type ChangeOrderStatus = "draft" | "submitted" | "approved" | "rejected";
export type ChangeOrderInitiator =
  | "client"
  | "contractor"
  | "designer"
  /** Auto-emitted by COSTPULSE when an approved estimate's total deviates
   *  >2% from the project's prior approved baseline. The accompanying
   *  `ai_analysis` carries `{ source: "costpulse.estimate_approved",
   *  prior_estimate_id, new_estimate_id, variance_pct, ... }`. */
  | "costpulse";
export type ReportStatus = "draft" | "sent" | "archived";
export type RAG = "green" | "amber" | "red";

export interface Task {
  id: UUID;
  organization_id: UUID;
  project_id: UUID;
  parent_id: UUID | null;
  title: string;
  description: string | null;
  status: TaskStatus;
  priority: Priority;
  assignee_id: UUID | null;
  phase: Phase | null;
  discipline: string | null;
  start_date: ISODate | null;
  due_date: ISODate | null;
  completed_at: string | null;
  position: number | null;
  tags: string[];
  created_by: UUID | null;
  created_at: string;
}

export interface TaskCreate {
  project_id: UUID;
  title: string;
  description?: string | null;
  status?: TaskStatus;
  priority?: Priority;
  assignee_id?: UUID | null;
  phase?: Phase | null;
  discipline?: string | null;
  start_date?: ISODate | null;
  due_date?: ISODate | null;
  position?: number | null;
  tags?: string[];
  parent_id?: UUID | null;
}

export type TaskUpdate = Partial<Omit<TaskCreate, "project_id">>;

export interface TaskBulkItem {
  id: UUID;
  status?: TaskStatus;
  phase?: Phase;
  position?: number;
  assignee_id?: UUID | null;
}

export interface TaskBulkUpdate {
  items: TaskBulkItem[];
}

export interface Milestone {
  id: UUID;
  organization_id: UUID;
  project_id: UUID;
  name: string;
  due_date: ISODate;
  status: MilestoneStatus;
  achieved_at: string | null;
}

export interface ChangeOrderAIAnalysis {
  root_cause: "design_change" | "scope_creep" | "site_condition" | "error" | "other";
  cost_breakdown: Record<string, unknown>;
  schedule_analysis: Record<string, unknown>;
  contract_clauses: string[];
  recommendation: "approve" | "negotiate" | "reject" | "request_more_info";
  reasoning: string;
  confidence: number;
}

/** Provenance payload for change orders auto-emitted by COSTPULSE when an
 *  approved estimate's total deviates from the project's prior approved
 *  baseline. Discriminated from the AI-analysis shape via `source`. */
export interface CostpulseVarianceAnalysis {
  source: "costpulse.estimate_approved";
  prior_estimate_id: UUID;
  new_estimate_id: UUID;
  prior_total_vnd: number;
  new_total_vnd: number;
  delta_vnd: number;
  variance_pct: number;
}

export type ChangeOrderAnalysis =
  | ChangeOrderAIAnalysis
  | CostpulseVarianceAnalysis;

export function isCostpulseVariance(
  a: ChangeOrderAnalysis | null | undefined,
): a is CostpulseVarianceAnalysis {
  return !!a && (a as CostpulseVarianceAnalysis).source === "costpulse.estimate_approved";
}

export interface ChangeOrder {
  id: UUID;
  organization_id: UUID;
  project_id: UUID;
  number: string;
  title: string;
  description: string | null;
  status: ChangeOrderStatus;
  initiator: ChangeOrderInitiator | null;
  cost_impact_vnd: number | null;
  schedule_impact_days: number | null;
  ai_analysis: ChangeOrderAnalysis | null;
  submitted_at: string | null;
  approved_at: string | null;
  approved_by: UUID | null;
  created_at: string;
}

export interface ChangeOrderCreate {
  project_id: UUID;
  number: string;
  title: string;
  description?: string | null;
  initiator?: ChangeOrderInitiator | null;
  cost_impact_vnd?: number | null;
  schedule_impact_days?: number | null;
}

export interface ChangeOrderApproval {
  decision: "approve" | "reject";
  notes?: string;
}

export interface ActionItem {
  title: string;
  owner: string | null;
  owner_user_id: UUID | null;
  deadline: ISODate | null;
}

export interface MeetingStructured {
  summary: string;
  decisions: string[];
  action_items: ActionItem[];
  risks: string[];
  next_meeting: ISODate | null;
}

export interface MeetingNoteCreate {
  project_id: UUID;
  meeting_date: ISODate;
  attendees: string[];
  raw_notes: string;
}

export interface MeetingNote {
  id: UUID;
  organization_id: UUID;
  project_id: UUID;
  meeting_date: ISODate;
  attendees: string[];
  raw_notes: string | null;
  ai_structured: MeetingStructured | null;
  created_by: UUID | null;
  created_at: string;
}

export interface MeetingStructureRequest {
  raw_notes: string;
  language?: "vi" | "en";
  project_id?: UUID;
  meeting_note_id?: UUID;
  persist?: boolean;
}

export interface ClientReportContent {
  header_summary: string;
  progress_section: {
    narrative?: string;
    highlights?: string[];
    progress_pct?: number;
  };
  photos_section: Array<{ url: string; caption?: string }>;
  financials: {
    narrative?: string;
    summary?: {
      spent?: number | null;
      budget?: number | null;
      variance?: number | null;
    };
  } | null;
  issues: Array<{ title: string; status: string; impact: string }>;
  next_steps: string[];
}

export interface ClientReport {
  id: UUID;
  organization_id: UUID;
  project_id: UUID;
  report_date: ISODate;
  period: string | null;
  content: ClientReportContent | null;
  rendered_html: string | null;
  pdf_url: string | null;
  status: ReportStatus;
  sent_at: string | null;
  sent_to: string[] | null;
}

export interface ReportGenerateRequest {
  project_id: UUID;
  period: string;
  date_from?: ISODate;
  date_to?: ISODate;
  language?: "vi" | "en";
  include_photos?: boolean;
  include_financials?: boolean;
}

export interface ReportSendRequest {
  recipients: string[];
  subject?: string;
  message?: string;
}

export interface TaskCountsByStatus {
  todo: number;
  in_progress: number;
  review: number;
  done: number;
  blocked: number;
}

export interface ProjectDashboard {
  project_id: UUID;
  rag_status: RAG;
  progress_pct: number;
  task_counts: TaskCountsByStatus;
  overdue_tasks: number;
  upcoming_milestones: Milestone[];
  open_change_orders: number;
  open_cost_impact_vnd: number;
  last_report_date: ISODate | null;
  alerts: string[];
}

File: apps/api/schemas/pulse.py
"""Pydantic schemas for PROJECTPULSE module."""

from __future__ import annotations

from datetime import date, datetime
from enum import StrEnum
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

# ---------- Enums ----------


class TaskStatus(StrEnum):
    todo = "todo"
    in_progress = "in_progress"
    review = "review"
    done = "done"
    blocked = "blocked"


class Priority(StrEnum):
    low = "low"
    normal = "normal"
    high = "high"
    urgent = "urgent"


class Phase(StrEnum):
    design = "design"
    permit = "permit"
    construction = "construction"
    closeout = "closeout"


class MilestoneStatus(StrEnum):
    upcoming = "upcoming"
    achieved = "achieved"
    missed = "missed"


class ChangeOrderStatus(StrEnum):
    draft = "draft"
    submitted = "submitted"
    approved = "approved"
    rejected = "rejected"


class ChangeOrderInitiator(StrEnum):
    client = "client"
    contractor = "contractor"
    designer = "designer"


class ReportStatus(StrEnum):
    draft = "draft"
    sent = "sent"
    archived = "archived"


class RAG(StrEnum):
    green = "green"
    amber = "amber"
    red = "red"


# ---------- Tasks ----------


class TaskBase(BaseModel):
    title: str = Field(min_length=1, max_length=500)
    description: str | None = None
    status: TaskStatus = TaskStatus.todo
    priority: Priority = Priority.normal
    assignee_id: UUID | None = None
    phase: Phase | None = None
    discipline: str | None = None
    start_date: date | None = None
    due_date: date | None = None
    position: float | None = None
    tags: list[str] = Field(default_factory=list)
    parent_id: UUID | None = None


class TaskCreate(TaskBase):
    project_id: UUID


class TaskUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=500)
    description: str | None = None
    status: TaskStatus | None = None
    priority: Priority | None = None
    assignee_id: UUID | None = None
    phase: Phase | None = None
    discipline: str | None = None
    start_date: date | None = None
    due_date: date | None = None
    position: float | None = None
    tags: list[str] | None = None
    parent_id: UUID | None = None


class Task(TaskBase):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    project_id: UUID
    organization_id: UUID
    completed_at: datetime | None = None
    created_by: UUID | None = None
    created_at: datetime


class TaskBulkItem(BaseModel):
    id: UUID
    status: TaskStatus | None = None
    phase: Phase | None = None
    position: float | None = None
    assignee_id: UUID | None = None


class TaskBulkUpdate(BaseModel):
    items: list[TaskBulkItem] = Field(min_length=1, max_length=200)


class TaskListFilters(BaseModel):
    project_id: UUID | None = None
    assignee_id: UUID | None = None
    phase: Phase | None = None
    status: TaskStatus | None = None
    parent_id: UUID | None = None
    limit: int = Field(default=50, ge=1, le=200)
    offset: int = Field(default=0, ge=0)


# ---------- Milestones ----------


class MilestoneBase(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    due_date: date
    status: MilestoneStatus = MilestoneStatus.upcoming


class MilestoneCreate(MilestoneBase):
    project_id: UUID


class Milestone(MilestoneBase):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    project_id: UUID
    organization_id: UUID
    achieved_at: datetime | None = None


# ---------- Change orders ----------


class ChangeOrderBase(BaseModel):
    number: str = Field(min_length=1, max_length=50)
    title: str = Field(min_length=1, max_length=500)
    description: str | None = None
    initiator: ChangeOrderInitiator | None = None
    cost_impact_vnd: int | None = None
    schedule_impact_days: int | None = None


class ChangeOrderCreate(ChangeOrderBase):
    project_id: UUID


class ChangeOrderUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    status: ChangeOrderStatus | None = None
    initiator: ChangeOrderInitiator | None = None
    cost_impact_vnd: int | None = None
    schedule_impact_days: int | None = None


class ChangeOrderAIAnalysis(BaseModel):
    root_cause: Literal["design_change", "scope_creep", "site_condition", "error", "other"]
    cost_breakdown: dict[str, Any] = Field(default_factory=dict)
    schedule_analysis: dict[str, Any] = Field(default_factory=dict)
    contract_clauses: list[str] = Field(default_factory=list)
    recommendation: Literal["approve", "negotiate", "reject", "request_more_info"]
    reasoning: str
    confidence: float = Field(ge=0.0, le=1.0)


class ChangeOrder(ChangeOrderBase):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    project_id: UUID
    organization_id: UUID
    status: ChangeOrderStatus
    ai_analysis: ChangeOrderAIAnalysis | None = None
    submitted_at: datetime | None = None
    approved_at: datetime | None = None
    approved_by: UUID | None = None
    created_at: datetime


class ChangeOrderApproval(BaseModel):
    decision: Literal["approve", "reject"]
    notes: str | None = None


class ChangeOrderListFilters(BaseModel):
    project_id: UUID | None = None
    status: ChangeOrderStatus | None = None
    limit: int = Field(default=50, ge=1, le=200)
    offset: int = Field(default=0, ge=0)


# ---------- Meeting notes ----------


class ActionItem(BaseModel):
    title: str
    owner: str | None = None
    owner_user_id: UUID | None = None
    deadline: date | None = None


class MeetingStructured(BaseModel):
    summary: str
    decisions: list[str] = Field(default_factory=list)
    action_items: list[ActionItem] = Field(default_factory=list)
    risks: list[str] = Field(default_factory=list)
    next_meeting: date | None = None


class MeetingNoteCreate(BaseModel):
    project_id: UUID
    meeting_date: date
    attendees: list[str] = Field(default_factory=list)
    raw_notes: str = Field(min_length=1)


class MeetingNote(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    project_id: UUID
    organization_id: UUID
    meeting_date: date
    attendees: list[str] = Field(default_factory=list)
    raw_notes: str | None = None
    ai_structured: MeetingStructured | None = None
    created_by: UUID | None = None
    created_at: datetime


class MeetingStructureRequest(BaseModel):
    raw_notes: str = Field(min_length=1, max_length=20000)
    language: Literal["vi", "en"] | None = None
    project_id: UUID | None = None
    meeting_note_id: UUID | None = None
    persist: bool = True


# ---------- Client reports ----------


class ReportGenerateRequest(BaseModel):
    project_id: UUID
    period: str = Field(description="e.g. 2026-W16, 2026-04, custom")
    date_from: date | None = None
    date_to: date | None = None
    language: Literal["vi", "en"] = "vi"
    include_photos: bool = True
    include_financials: bool = True


class ClientReportContent(BaseModel):
    header_summary: str
    progress_section: dict[str, Any]
    photos_section: list[dict[str, Any]] = Field(default_factory=list)
    financials: dict[str, Any] | None = None
    issues: list[dict[str, Any]] = Field(default_factory=list)
    next_steps: list[str] = Field(default_factory=list)


class ClientReport(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    project_id: UUID
    organization_id: UUID
    report_date: date
    period: str | None = None
    content: ClientReportContent | None = None
    rendered_html: str | None = None
    pdf_url: str | None = None
    status: ReportStatus
    sent_at: datetime | None = None
    sent_to: list[str] | None = None


class ReportSendRequest(BaseModel):
    recipients: list[str] = Field(min_length=1)
    subject: str | None = None
    message: str | None = None


# ---------- Dashboard ----------


class TaskCountsByStatus(BaseModel):
    todo: int = 0
    in_progress: int = 0
    review: int = 0
    done: int = 0
    blocked: int = 0


class ProjectDashboard(BaseModel):
    project_id: UUID
    rag_status: RAG
    progress_pct: float = Field(ge=0.0, le=100.0)
    task_counts: TaskCountsByStatus
    overdue_tasks: int
    upcoming_milestones: list[Milestone]
    open_change_orders: int
    open_cost_impact_vnd: int
    last_report_date: date | None = None
    alerts: list[str] = Field(default_factory=list)
